phac-nml/irida-next
# Project: Bot Accounts

## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR adds in bot accounts for projects which can be used to perform actions which can't be attributed to a human user for example using an uploader.

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._

![image](https://github.com/phac-nml/irida-next/assets/25466211/293c024d-ba20-4bf2-afc1-1e86bbacf46e)

![image](https://github.com/phac-nml/irida-next/assets/25466211/23a12ecc-e308-4197-abcf-ec6b3837f9ba)

![image](https://github.com/phac-nml/irida-next/assets/25466211/5cd4dda7-241f-4a7f-9424-b07017552ba2)

![image](https://github.com/phac-nml/irida-next/assets/25466211/e40e7aee-b270-4d8f-9400-f07f5612c6e5)

**Member added automatically when project bot account is created**

![image](https://github.com/phac-nml/irida-next/assets/25466211/0a84d713-0217-4e81-ad16-3805b7c25441)


## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._

1. Login as any user
2. Create a project
3. Go to the show page for the project
4. From the left side bar select `Settings` and from the dropdown menu select `Bot Accounts`
5. Verify you see an empty state message for no bot accounts
6. Click the `Add New Bot Account` button
7. In the `New Bot Account` modal, fill in the `Token Name`, select an access level, set an optional expiry date, and select atleast one of the scopes, and then click the `Submit` button
8. You should now see a listing for bot accounts with 1 row for the bot account that you created, and above the table you should see a section for the personal access token which the user needs to copy as it cannot be accessed afterwards
9. Verify the bot account is added as a member to the project from the `Members` page
10. Verify you can delete a bot account by visiting the `Bot Accounts` page for the project and then clicking the ellipsis menu for a bot account and selecting `Remove` and then `Confirm` in the confirmation modal.
11. Verify if you have bot accounts for a project but they aren't members (membership removed from the `Members` page), and the project is deleted, that the project bot user accounts are deleted as well. This can be done via the rails console and inspecting the `users` table

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.
